File: app/src/main/java/com/holokenmod/creation/GridCageOperationDecider.java
package com.holokenmod.creation;

import com.holokenmod.GridCage;
import com.holokenmod.GridCageAction;
import com.holokenmod.RandomSingleton;
import com.holokenmod.options.GridCageOperation;

import java.util.Optional;

class GridCageOperationDecider {
	private final GridCage cage;
	private final GridCageOperation operationSet;
	
	GridCageOperationDecider(GridCage cage, GridCageOperation operationSet) {
		this.cage = cage;
		this.operationSet = operationSet;
	}
	
	Optional<GridCageAction> decideOperation() {
		if (cage.getCells().size() == 1) {
			return Optional.empty();
		}
		
		if (operationSet == GridCageOperation.OPERATIONS_MULT) {
			return Optional.of(GridCageAction.ACTION_MULTIPLY);
		}
		
		CageCalculationDecision decision = getCalculationDecision(operationSet);
		
		switch (decision) {
			case ADDITION_AND_SUBTRACTION:
				return Optional.of(decideBetweenAdditionAndSubtraction(cage, operationSet));
			case MULTIPLICATION_AND_DIVISION:
				return Optional.of(decideBetweenMultiplicationAndDivision(cage, operationSet));
		}
		
		throw new RuntimeException("Decision could not be calculated.");
	}
	
	private GridCageAction decideBetweenAdditionAndSubtraction(GridCage cage, GridCageOperation operationSet) {
		if (cage.getCells().size() > 2 || operationSet == GridCageOperation.OPERATIONS_ADD_MULT) {
			return GridCageAction.ACTION_ADD;
		}
		
		final double randomValue = RandomSingleton.getInstance().nextDouble();
		
		//if (randomValue >= 0.25) {
		//	return GridCageAction.ACTION_SUBTRACT;
		//}
		
		return GridCageAction.ACTION_ADD;
	}
	
	private GridCageAction decideBetweenMultiplicationAndDivision(GridCage cage, GridCageOperation operationSet) {
		if (cage.getCells().size() > 2 || operationSet == GridCageOperation.OPERATIONS_ADD_MULT) {
			return GridCageAction.ACTION_MULTIPLY;
		}
		
		final double randomValue = RandomSingleton.getInstance().nextDouble();
		
		if (randomValue >= 0.25 && cage.canHandleDivide()) {
			return GridCageAction.ACTION_DIVIDE;
		}
		
		return GridCageAction.ACTION_MULTIPLY;
	}
	
	private CageCalculationDecision getCalculationDecision(GridCageOperation operationSet) {
		if (operationSet == GridCageOperation.OPERATIONS_ADD_SUB) {
			return CageCalculationDecision.ADDITION_AND_SUBTRACTION;
		}
		
		final double randomValue = RandomSingleton.getInstance().nextDouble();
		
		if (randomValue >= 0.5) {
			return CageCalculationDecision.MULTIPLICATION_AND_DIVISION;
		} else {
			return CageCalculationDecision.ADDITION_AND_SUBTRACTION;
		}
	}
}

File: holoken-core/src/main/kotlin/com/holokenmod/creation/cage/GridCageOperationDecider.kt
package com.holokenmod.creation.cage

import com.holokenmod.Randomizer
import com.holokenmod.grid.GridCageAction
import com.holokenmod.grid.GridCell
import com.holokenmod.options.GridCageOperation
import kotlin.math.max
import kotlin.math.min

internal class GridCageOperationDecider(
    private val randomizer: Randomizer,
    private val cells: List<GridCell>,
    private val operationSet: GridCageOperation
) {
    fun decideOperation(): GridCageAction? {
        if (cells.size == 1) {
            return null
        }
        if (operationSet == GridCageOperation.OPERATIONS_MULT) {
            return GridCageAction.ACTION_MULTIPLY
        }
        return when (getCalculationDecision()) {
            CageCalculationDecision.ADDITION_AND_SUBTRACTION ->
                decideBetweenAdditionAndSubtraction()
            CageCalculationDecision.MULTIPLICATION_AND_DIVISION ->
                decideBetweenMultiplicationAndDivision()
        }
    }

    private fun decideBetweenAdditionAndSubtraction(): GridCageAction {
        if (cells.size > 2 || operationSet == GridCageOperation.OPERATIONS_ADD_MULT) {
            return GridCageAction.ACTION_ADD
        }
        val randomValue = randomizer.nextDouble()
        return if (randomValue >= 0.25) {
            GridCageAction.ACTION_SUBTRACT
        } else {
            GridCageAction.ACTION_ADD
        }
    }

    private fun decideBetweenMultiplicationAndDivision(): GridCageAction {
        if (cells.size > 2 || operationSet == GridCageOperation.OPERATIONS_ADD_MULT) {
            return GridCageAction.ACTION_MULTIPLY
        }
        val randomValue = randomizer.nextDouble()
        return if (randomValue >= 0.25 && canHandleDivide()) {
            GridCageAction.ACTION_DIVIDE
        } else {
            GridCageAction.ACTION_MULTIPLY
        }
    }

    private fun canHandleDivide(): Boolean {
        val cell1Value = cells[0].value
        val cell2Value = cells[1].value

        val higher = max(cell1Value, cell2Value)
        val lower = min(cell1Value, cell2Value)

        if (lower == 0 && higher == 0) {
            return false
        }

        return if (lower == 0 || higher == 0) {
            true
        } else {
            higher % lower == 0
        }
    }

    private fun getCalculationDecision(): CageCalculationDecision {
        if (operationSet == GridCageOperation.OPERATIONS_ADD_SUB) {
            return CageCalculationDecision.ADDITION_AND_SUBTRACTION
        }
        val randomValue = randomizer.nextDouble()
        return if (randomValue >= 0.5) {
            CageCalculationDecision.MULTIPLICATION_AND_DIVISION
        } else {
            CageCalculationDecision.ADDITION_AND_SUBTRACTION
        }
    }
}
